src/librustc_mir/borrow_check/nll/region_infer/mod.rs

```diff
@@ -465,37 +466,58 @@ impl<'tcx> RegionInferenceContext<'tcx> {
         // constraints we have accumulated.
         let mut inferred_values = self.liveness_constraints.clone();
 
-        while changed {
-            changed = false;
-            debug!("propagate_constraints: --------------------");
-            for constraint in &self.constraints {
-                debug!("propagate_constraints: constraint={:?}", constraint);
-
-                // Grow the value as needed to accommodate the
-                // outlives constraint.
-                let Ok(made_changes) = self.dfs(
-                    mir,
-                    CopyFromSourceToTarget {
-                        source_region: constraint.sub,
-                        target_region: constraint.sup,
-                        inferred_values: &mut inferred_values,
-                        constraint_point: constraint.point,
-                        constraint_span: constraint.span,
-                    },
-                );
+        let dependency_map = self.build_dependency_map();
+        let mut dirty_list: Vec<_> = (0..self.constraints.len()).collect();
+        let mut dirty_bit_vec = BitVector::new(dirty_list.len());
+
+        debug!("propagate_constraints: --------------------");
+        while let Some(constraint_idx) = dirty_list.pop() {
+            dirty_bit_vec.remove(constraint_idx);
+
+            let constraint = &self.constraints[constraint_idx];
+            debug!("propagate_constraints: constraint={:?}", constraint);
+
+            // Grow the value as needed to accommodate the
+            // outlives constraint.
+            let Ok(made_changes) = self.dfs(
+                mir,
+                CopyFromSourceToTarget {
+                    source_region: constraint.sub,
+                    target_region: constraint.sup,
+                    inferred_values: &mut inferred_values,
+                    constraint_point: constraint.point,
+                    constraint_span: constraint.span,
+                },
+            );
+
+            if made_changes {
+                debug!("propagate_constraints:   sub={:?}", constraint.sub);
+                debug!("propagate_constraints:   sup={:?}", constraint.sup);
 
-                if made_changes {
-                    debug!("propagate_constraints:   sub={:?}", constraint.sub);
-                    debug!("propagate_constraints:   sup={:?}", constraint.sup);
-                    changed = true;
+                for dep_idx in dependency_map.get(&constraint.sup).unwrap_or(&vec![]) {
+                    if !dirty_bit_vec.contains(*dep_idx) {
+                        dirty_bit_vec.insert(*dep_idx);
+                        dirty_list.push(*dep_idx);
+                    }
                 }
             }
+
             debug!("\n");
         }
 
         self.inferred_values = Some(inferred_values);
     }
 
+    fn build_dependency_map(&self) -> HashMap<RegionVid, Vec<usize>> {
```

Comment:
Maybe a comment here would be nice, actually. You could steal the stuff I wrote in the issue description (i.e., the example). But something like:

---

/// Builds up a map from each region variable X to a vector with the indices of constraints that
/// need to be re-evaluated when X changes. These are constraints like `Y: X @ P` -- so if `X`
/// changed, we may need to grow Y.